### PR TITLE
Hook improvements

### PR DESCRIPTION
**Pull request by dramalho**
This PR changes the way Hooks are registered and handled in the following way:

* Hook handlers can now be arbitrarily registered and multiple handlers can be defined for the same event type
* Anything that implements `call` (from lambdas to custom objects with the call method) can be registered as a hook.
* `SlackRubyBot::Hooks::Base` is gone, Hooks no longer need to extend this class and they will no longer be implicitly registered this way
* `SlackRubyBot::Server` has one instance of `SlackRubyBot::Hooks:Set` which handles handler registration. Configuration is handled by an optional `hook_handlers` setting

Missing:

* [x] Upgrade instructions
* [x] Command like hook blocks

````ruby
  hook 'event' do |client, data|
    # Do something
  end
````

### CODE REVIEW

File: lib/slack-ruby-bot/hooks/set.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      def initialize(client = nil)
+        self.handlers = {}
+        self.client = client
+
+        @pending_flush = client.blank?
+      end
+
+      def add(hook_name, handler)
+        unless handlers[hook_name].present?
+          handlers[hook_name] ||= []
```
Comment (dblock) [inline lib/slack-ruby-bot/hooks/set.rb]: No big deal ... This can be a straight assignment because of `unless`.

**Comment by dblock** [comment]
This looks great to me. Needs CHANGELOG and README updates and squashed commits and it's good to go.

File: lib/slack-ruby-bot/hooks/set.rb (hunk 1 above)
Comment (dramalho) [inline lib/slack-ruby-bot/hooks/set.rb]: correct! :)

**Comment by dramalho** [comment]
Ok, just wanted to add support for hook commands and I ended up moving some of the hook set stuff into a module.

````ruby
hook 'hello' do |_ = nil, _ = nil|
   'hello'
end
````

Tomorrow I'll take a look at README and UPGRADING and squashing commits

**Comment by dblock** [comment]
Looking forward to it!

**Comment by dramalho** [comment]
@dblock ok, took me long enough, the Weekend didn't turn out as expected :)

It's mostly there, can you take a good look before I squash the commits and all that jazz?

Thanks Daniel :) sorry for the slow turnaround , busy times :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,11 @@
-### 0.7.1 (Next)
+### 0.8.1 (Next)
```
Comment (dblock) [inline CHANGELOG.md]: We want 0.8.0 to be next since there're interface changes, so just change 0.7.1 to 0.8.0, no need to have a 0.8.1 here.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,11 @@
...
 
 * Your contribution here.
 
+### 0.8.0 (Next)
+
+* [#54](https://github.com/dblock/slack-ruby-bot/pull/54) Improvements to Hook configuration - [@dramalho](https://github.com/dramalho)
```
Comment (dblock) [inline CHANGELOG.md]: Missing a period at the end.

File: README.md [hunk 4]
```diff
@@ -190,17 +190,36 @@ Get help.
...
+```ruby
+SlackRubyBot::Server.new(hook_handlers: {
+  hello: MyBot::Hooks::UserChange.new
+})
+```
+
+or at any time by pushing it to the `HookSet`
+
+```ruby
+server.hook_set.add(:hello, MyBot::Hooks::UserChange.new)
```
Comment (dblock) [inline README.md]: I think `hook_set` should be `hooks`.

File: README.md [hunk 5]
```diff
@@ -210,6 +229,23 @@ module MyBot
 end
 ```
 
+Hooks can also be written as blocks inside the `SlackBotRuby::Server` class, for example
+
+```ruby
+module MyBot
+  class MyServer < SlackRubyBot::Server
+    hook 'hello' do |client, data|
```
Comment (dblock) [inline README.md]: I wonder whether `hook` here should be `on`?

File: UPGRADING.md [hunk 6]
```diff
@@ -1,6 +1,16 @@
 Upgrading SlackRubyBot
 ======================
 
+### Upgrading to >= 0.8.0
+
+#### Hook::Base was removed, explicitly register any custom hook classes you have
```
Comment (dblock) [inline UPGRADING.md]: I would remove "you have"

File: CHANGELOG.md (hunk 2 above)
Comment (dramalho) [inline CHANGELOG.md]: ok, I went over the guidelines and because this is a breaking change I bumped to 0.8, that was why. I'll put it back

File: UPGRADING.md [hunk 7]
```diff
@@ -1,6 +1,16 @@
...
+
+Hook classes are now handled differently, namely they are explicitly registered into `SlackRubyBot::Server` via a configuration option (`hook_handlers`) or by passing a similar hash later on through the `add_hook_handlers` method. Including Hook classes directly into the server class is no longer needed.
+
+A hook is actioned via a `call` message onto the handler object (this can be anything that responds to that), so you'll need to rename your method names.
```
Comment (dblock) [inline UPGRADING.md]: rename your methods ... rename your method names is a bit like "buttered buttered sandwich" in French :)

File: README.md (hunk 4 above)
Comment (dramalho) [inline README.md]: :)

sure

File: UPGRADING.md [hunk 8]
```diff
@@ -1,6 +1,16 @@
...
+
+Finally, you can now register multiple hooks for the same event, so if you had any code to remove default hooks, you'll need to change it so you pass a configuration hash into `Server`
```
Comment (dblock) [inline UPGRADING.md]: Provide an example of before/after, please.

File: lib/slack-ruby-bot/hooks/hello.rb [hunk 9]
```diff
@@ -1,9 +1,13 @@
 module SlackRubyBot
   module Hooks
-    module Hello
-      extend Base
+    class Hello
+      attr_accessor :logger
 
-      def hello(client, _data)
+      def initialize(logger)
```
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hello.rb]: The argument should be `options`, a Hash, so we can evolve this in the future without changing the interface.

File: README.md (hunk 5 above)
Comment (dramalho) [inline README.md]: I kept it explicit like the commands, I get that it's an event, but `on` feels way to short and generic. 

Up to you, happy to comply :)

File: lib/slack-ruby-bot/hooks/hook_support.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+module SlackRubyBot
+  module Hooks
+    module HookSupport
+      def self.included(base)
+        base.cattr_accessor :hook_blocks
```
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: I don't see a problem with calling both `hooks` on class and instance.

File: lib/slack-ruby-bot/hooks/hello.rb (hunk 9 above)
Comment (dramalho) [inline lib/slack-ruby-bot/hooks/hello.rb]: well, the initialiser is not part of the interface, this particular hook takes the logger (I didn't want to touch any of the existing implementations) but it's 100% irrelevant if a hook is an object / class / lambda / whatever, so it's really irrelevant how it gets initialised

Not sure if you follow me :)

File: lib/slack-ruby-bot/hooks/hook_support.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+      def flush_hook_blocks
+        return nil unless self.class.hook_blocks
+
+        add_hook_handlers(self.class.hook_blocks)
+      end
+
+      def add_hook_handlers(handler_hash)
+        handler_hash.each do |hook, handlers|
+          handlers = [handlers] unless handlers.is_a?(Enumerable)
```
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: I think you can do `handlers = Array(handlers)` here.

File: lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb [hunk 12]
```diff
@@ -2,12 +2,20 @@
 
 RSpec::Matchers.define :respond_with_error do |error, error_message|
   match do |actual|
-    channel, user, message = parse(actual)
     allow(Giphy).to receive(:random)
+
+    client = if respond_to?(:client)
```
Comment (dblock) [inline lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb]: Maybe a `1-liner`, `respond_to?(:client) ? send(:client) : SlackRubyBot::Client.new`.

**Comment by dblock** [comment]
This looks great, you should make the updates that you see fit, get it squashed and I'll merge, see comments.

File: CHANGELOG.md (hunk 2 above)
Comment (dblock) [inline CHANGELOG.md]: Right, just no need for 2 "next" versions.

File: README.md (hunk 5 above)
Comment (dblock) [inline README.md]: I think `on` is better, thanks.

File: lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb (hunk 12 above)
Comment (dramalho) [inline lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb]: I had a hard rubocop assisted transition away from one liners, don't break my therapy man :)

tl;dr; this just reads nicer, after a while I've started to agree

File: lib/slack-ruby-bot/hooks/hello.rb (hunk 9 above)
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hello.rb]: I follow. You can leave it as is.

File: lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb (hunk 12 above)
Comment (dblock) [inline lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_error.rb]: Lol.

File: CHANGELOG.md (hunk 2 above)
Comment (dramalho) [inline CHANGELOG.md]: yeah, my bad there. So to clarify, do I go to 0.7.1 or 0.8.0 on this one ?

File: README.md (hunk 5 above)
Comment (dramalho) [inline README.md]: :+1:

File: lib/slack-ruby-bot/hooks/hook_support.rb (hunk 11 above)
Comment (dramalho) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: right you are

File: CHANGELOG.md (hunk 2 above)
Comment (dblock) [inline CHANGELOG.md]: 0.8.0

Comment (dramalho) [inline CHANGELOG.md]: awesomesauce

File: UPGRADING.md (hunk 7 above)
Comment (dramalho) [inline UPGRADING.md]: buerré, I learned some French today :)

File: UPGRADING.md (hunk 8 above)
Comment (dramalho) [inline UPGRADING.md]: Not sure if there was a canonical approach, you could maybe manipulate the `hooks` attribute, but I'm not sure people would do that. That was actually what triggered this (https://github.com/dblock/slack-ruby-bot/issues/50) .

Would that be a fair example you think?

File: lib/slack-ruby-bot/hooks/hook_support.rb (hunk 10 above)
Comment (dramalho) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: in this case, it's just a collection of lambdas to hold the `on 'event_name'` execution blocks, once the class gets instantiated this gets flushed to the `Hook::Set` . Not that I'm super happy about that because we explicitly need to call the flush method.

thoughts?

**Comment by dramalho** [comment]
All squashed up!

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,6 +1,6 @@
-### 0.7.1 (Next)
+### 0.8.0
 
-* Your contribution here.
```
Comment (dblock) [inline CHANGELOG.md]: Put this line back please.

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,6 +1,6 @@
...
+* [#54](https://github.com/dblock/slack-ruby-bot/pull/54) Improvements to Hook configuration - [@dramalho](https://github.com/dramalho).
```
Comment (dblock) [inline CHANGELOG.md]: There's usually a `:` after the issue number/link.

File: README.md [hunk 15]
```diff
@@ -190,17 +190,40 @@ Get help.
 
 ### Hooks
 
-Hooks are event handlers and respond to Slack RTM API [events](https://api.slack.com/events), such as [hello](lib/slack-ruby-bot/hooks/hello.rb) or [message](lib/slack-ruby-bot/hooks/message.rb). You can implement your own by extending [SlackRubyBot::Hooks::Base](lib/slack-ruby-bot/hooks/base.rb).
+Hooks are event handlers and respond to Slack RTM API [events](https://api.slack.com/events), such as [hello](lib/slack-ruby-bot/hooks/hello.rb) or [message](lib/slack-ruby-bot/hooks/message.rb). You can implement your own in a couple of ways:
+
+
+#### Implement and register a Hook Handler
```
Comment (dblock) [inline README.md]: If we're going to say Hook Handler we should make Register capitalized probably. Or the other way around.

Comment (dramalho) [inline README.md]: hummmmm I capitalise concepts, not so much common actions (verbs) . But that's me :) , pick one (capitalise all or none), I'll change it accordingly

File: lib/slack-ruby-bot/hooks/hook_support.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        def on(event_name, &block)
+          self.hook_blocks ||= {}
+
+          self.hook_blocks[event_name] ||= []
+          self.hook_blocks[event_name] << block
+        end
+      end
+
+      # Instance stuff
+      attr_writer :hooks
```
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: Is this ever assigned? Maybe it should be an `attr_reader` and marked `private`?

File: lib/slack-ruby-bot/hooks/hook_support.rb (hunk 10 above)
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: Lets leave it as is.

File: README.md (hunk 15 above)
Comment (dblock) [inline README.md]: It's fine.

File: lib/slack-ruby-bot/hooks/hook_support.rb (hunk 16 above)
Comment (dramalho) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: well https://github.com/dblock/slack-ruby-bot/pull/54/files#diff-04c6e90faac2675aa89e2176d2eec7d8R212

**Comment by dblock** [comment]
Rebase and squash (make sure there's 1 commit here) and I'll merge as is. This looks great.

File: lib/slack-ruby-bot/hooks/hook_support.rb (hunk 16 above)
Comment (dblock) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: Are you misusing `attr_writer`? That makes `hooks` itself assignable, you want `attr_reader` no?

Comment (dramalho) [inline lib/slack-ruby-bot/hooks/hook_support.rb]: sorry, you're correct. This was a previous iteration, I don't need either as I'm implementing a reader method anyway. 

of with it's head!

**Comment by dblock** [comment]
You still have a merge commit here, you should rebase this branch.

```
git checkout master
git pull upstream master
git checkout hook-improvements
git rebase master
git push origin hoook-improvements -f
```

**Comment by dramalho** [comment]
more like missing a rebase on my master , all done now!

**Comment by dramalho** [comment]
`13 days ago`, ffs , my handle on life is in need of reinforcement :)

**Comment by dblock** [comment]
Merged, thanks!

**Comment by dramalho** [comment]
:tada: :balloon:
